app/controllers/admin/delivery_partners_controller.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+class Admin::DeliveryPartnersController < Admin::BaseController
+  skip_after_action :verify_authorized
+  skip_after_action :verify_policy_scoped
+
+  def choose_lead_providers
+    authorize DeliveryPartner, :create?
+
+    new_supplier_form = NewSupplierForm.new(session[:new_supplier_form])
+    if session[:delivery_partner_form].nil?
+      session[:delivery_partner_form] = { name: new_supplier_form.name }
+    else
+      session[:delivery_partner_form].merge!({ name: new_supplier_form.name })
+    end
+
+    @delivery_partner_form = DeliveryPartnerForm.new(session[:delivery_partner_form])
+  end
+
+  def receive_lead_providers
+    authorize DeliveryPartner, :create?
+
+    provider_relationships = []
+    # Ensure all selected lead providers have at least one selected cohort
+    chosen_lead_providers = params.dig(:delivery_partner_form, :lead_providers)&.keep_if(&:present?)
+
+    unless chosen_lead_providers&.any?
+      @delivery_partner_form = DeliveryPartnerForm.new(session[:delivery_partner_form])
+      @delivery_partner_form.errors.add(:lead_providers, :blank, message: "Choose at least one")
+      render :choose_lead_providers and return
+    end
+
+    chosen_lead_providers.each do |lead_provider|
+      chosen_cohorts = params.dig(
+        :delivery_partner_form,
+        lead_provider.to_sym,
+        :cohorts,
+      )&.keep_if(&:present?)
+
+      unless chosen_cohorts&.any?
+        @delivery_partner_form = DeliveryPartnerForm.new(session[:delivery_partner_form])
+        @delivery_partner_form.errors.add(:lead_providers, :blank, message: "Choose at least one cohort for every selected lead provider")
+        # rubocop:disable Lint/NonLocalExitFromIterator
+        render :choose_lead_providers and return
+        # rubocop:enable Lint/NonLocalExitFromIterator
+      end
+
+      chosen_cohorts.each do |cohort|
+        provider_relationships.push ProviderRelationship.new(
+          cohort: Cohort.find(cohort),
+          lead_provider: LeadProvider.find(lead_provider),
+        )
+      end
```

Comment:
Some potential for N+1 queries here I guess, but screw it, we can optimise it when it becomes an issue (I think).